cosa/20200805-meta-merge.md

```diff
@@ -0,0 +1,103 @@
+# Merging of build meta-data
+- Specification ID: cosa-20200805
+- Supersedes: None
+- Conflicts: None
+
+# Background
+
+One of the biggest blockers to performing parallel builds, distributed builds and/or re-starting failed builds is the ability to lock write access to `meta.json` in COSA.
+
+Under the following assmptions, COSA needs to support merging of `meta.json`:
+- Not all `cmd-buildextend-*` operations will happen on the same pod or Linux namespace
+- S3 is the authoritative store for builds. S3 is eventually consistent meaning that the version read from S3 may be replaced after reading from S3. Implementing distributed locking on `meta.json` based on S3 is problematic.
+- For RHCOS we need the ability to do partial builds and then back-fill artifacts (e.g. only build qemu, but not cloud) and later build cloud.
+- Developers want to restart failed builds
+- Consumers of the artifacts from COSA want to reduce the time to build and test
+
+## State of a build is a point-of-view statement
+
+To COSA a build is usually `ostree` + `qcow2`. The notion of a "failed" or "partial" build is a point of view. COSA does provision the build id, but COSA does not know what comprised the artifacts of the "build". Usually developers and consumers speak of a build as the base `ostree` and the _optional_ artifacts. 
+
+In COSA a "build" only starts to exist when the `meta.json` is incepted after the `ostree` has been written. The build directory may exist, but the `meta.json` that is produced immediately is valid for any consumer. In most cases a "failed" build means that an artifact in the `buildextend-*` commands has failed to publish (such as AWS) not that the artifact itself failed to build; the artifact was indeed built. 
+
+Further, a "failed build" is also used to mean that some part of a pipeline failed to complete. COSA itself does not have any concept of dependencies, much less the idea that a multiple COSA commands can be run. COSA only has implied dependencies that some artifacts requires others. In other words, COSA doesn't know when the "build" is complete until after something else considers it's complete. 
+
+Denoting the state of a build in the `meta.json` will be owned by process/thing/person that is orchestrating the build. 
+
+## Multiple `meta.json`
+
+Under the axiom that the state of a build is usually determined from the point-of-view of a pipeline and to support distributed and parallel builds, builds will be produced with multiple `meta.json` that will be merged together.
```

Comment:
Interesting idea... I will explore that, but to use a stub I need to rework the design. I'm trying to think through the implications of what that would mean. The obvious implications get a little funny around the "when" since the default build mode would call it. Are you thinking that there would be `cmd-meta-merge --final` for the `coreos-assembler.delay-meta-merge`?

Reply:
I've updated the spec for a `stub.<artifact>.json`. But to be honest, I do not like introducing another data-model. The advantage of having complete `meta.json` is that it follows the schema and that means things like `kola` understand it. The other reason I like having the complete `meta.json.<ARTIFACT>` is that it gives the ability to ensure that the builds are the same. 

After thinking this through the only way I could get on board is to introduce another schema to ensure that the artifacts are from the same build -- we can compare the name, build id and ostree to ensure that its the same.

Relying on `jq -s add` may be simple for the bash case, but that won't work with the `cosalib.meta` case. After looking at the code, this is actually harder to implement for the Pythonic code. 

> That way, there is no room for misinterpretation or for multiple sources of truth, etc... and it makes the merging step dead simple.

Merging is dead simple for bash, not so much for Python. If, per the other comments, we are not publishing the artifacts JSON, then what will mis-interpret it? We could use a different name, e.g. `stub.<ARTIFACT>.json` to reflect its incomplete. Today the tooling only understands `builds.json` and `meta.json`. So the only case that I can think of is the developer case. 

I've updated the proposal to reflect this idea, but I'm not a fan.

Reply:
Ack, thanks for considering. I think following the schema is totally fine if it simplifies the implementation. If we're agreed on deleting the files, as you said there's nothing else that should consume it anyway. So going back to non-fragments SGTM!

Reply:
What do you think of using a naming convention of `artifacts.<ARTIFACT>.json`  (e.g. `artifacts.aws.json`)?

Reply:
I like that it ends with `.json` (nice for editors/highlighters). Another suggestion which is a slight tweak on your original proposal is `meta.<ARTIFACT>.json` so it stays close to `meta.json` in listings. WFM either way though.